src/main/java/com/machine/learning/classifier/ID3.java

```diff
@@ -0,0 +1,318 @@
+	    }
+	}
+	
+	return false;
+    }
+
+    private int validationError(DecisionTree dt) {
+	int errors = 0;
+	for (DataPoint point : validationData) {
+	    if (!classify(point.getData().get(), dt).equals(point.getClassLabel().get())) {
+		errors++;
+	    }
+	}
+	return errors;
+    }
+
+    @Override
+    public String classify(List dataPoint) {
+	return classify(dataPoint, dt);
+    }
+
+    private String classify(List dataPoint, DecisionTree curDT) {
+	while (curDT.clazz == null) {
+	    if (dataPoint.get(curDT.attributeIndex).equals(curDT.attributeValue)) {
+		curDT = curDT.pos;
+	    } else {
+		curDT = curDT.neg;
+	    }
+	}
+	
+	return curDT.clazz;
+    }
+
+    public DecisionTree constructDT(List<DataPoint> remainingData) {
+	if (remainingData == null || remainingData.size() == 0) {
+	    return null;
+	}
+	Set<String> classes = new HashSet<>();
+	List<Set<String>> usedAttrValues = new ArrayList<>();
+
+	for (DataPoint dataPoint : remainingData) {
+	    classes.add(dataPoint.getClassLabel().get());
+	}
+
+	if (classes.size() == 1) {
+	    String singleClass = classes.iterator().next();
+	    return new DecisionTree(singleClass);
+	}
+	
+
+	for(int i = 0; i < remainingData.get(0).getData().get().size(); i++) {
+	    usedAttrValues.add(new HashSet<String>());
+	}
+
+	for (DataPoint dataPoint : remainingData) {
+	    List<String> datum = dataPoint.getData().get();
+	    for (int i = 0; i < datum.size(); i++) { 
+		usedAttrValues.get(i).add(datum.get(i));
+	    }
+	}
+
+	int attrIndex = 0;
+	String attributeValue = "";
+	double minEntropy = Double.MAX_VALUE;
+	ArrayList<DataPoint> posData = new ArrayList<>();
+	ArrayList<DataPoint> negData = new ArrayList<>();
+	double curEntropy = calculateEntropy(remainingData);
+	for (int i = 0; i < usedAttrValues.size(); i++) {
+	    for (String attrValue : usedAttrValues.get(i)) {
+		posData.clear();
+		negData.clear();
+		
+		for (DataPoint dataPoint : remainingData) {
+		    if (dataPoint.getData().get().get(i).equals(attrValue)) {
+			posData.add(dataPoint);
+		    } else {
+			negData.add(dataPoint);
+		    }
+		}
+		
+		double entropy = (calculateEntropy(posData)*posData.size()/remainingData.size() +
+				  calculateEntropy(negData)*negData.size()/remainingData.size());
+		if(entropy < minEntropy){
+		    minEntropy = entropy;
+		    attrIndex = i;
+		    attributeValue = attrValue;
+		}		
+	    }
+	}
+
+
+	if (aboutEqual(curEntropy, minEntropy) || minEntropy > curEntropy) {
+	    return new DecisionTree(mostCommonClass(remainingData));
+	}
+	
+	posData.clear();
+	negData.clear();
+	
+	for (DataPoint dataPoint : remainingData) {
+	    if (dataPoint.getData().get().get(attrIndex).equals(attributeValue)) {
+		posData.add(dataPoint);
+	    } else {
+		negData.add(dataPoint);
+	    }
+	}
+	
+	DecisionTree cur = new DecisionTree(attrIndex, attributeValue);
+	cur.maxClass = mostCommonClass(remainingData);	
+	cur.pos = constructDT(posData);
+	cur.neg = constructDT(negData);
+
+	return cur;
+    }
```

Comment:
This method is super long and generally you would break this out. I won't make you though